src/Polly.Core/README.md

```diff
@@ -165,45 +165,105 @@ var resilienceStrategy = new ResilienceStrategyBuilder()
+- **`OutcomeGenerator<TArgs, TValue>`**: holds a list of generators for various result types. These generators produce the value of `TValue` type.
+- **`OutcomeGenerator<TArgs, TValue, TResult>`**: holds a generator for a specific result type.
+- **`VoidOutcomeGenerator<TArgs, TValue>`**: holds a generator for a void-based result type.
+
+API Usage:
+
 ``` csharp
 var options = new RetryStrategyOptions();
-options
-    .ShouldRetry
-    .Add<HttpResponseMessage>(result => result.StatusCode == HttpStatusCode.InternalServerError) // inspecting the result
-    .Add(HttpStatusCode.InternalServerError) // particular value for other type
-    .Add<MyResult>(result => result.IsError)
-    .Add<MyResult>((result, context) => IsError(context)) // retrieve data from context for evaluation
-    .AddException<InvalidOperationException>() // exceptions
-    .AddException<HttpRequestMessageException>() // more exceptions
-    .AddException(error => IsError(error)) // exception predicates
-    .Add<MyResult>(async (result, context) => await IsErrorAsync(result, context)); // async predicates
+OutcomeGenerator<RetryDelayArguments, TimeSpan> generator = options.RetryDelayGenerator;
+
+generator
+    .SetGenerator((outcome, args) => TimeSpan.FromSeconds(args.Attempt)) // called for all result types
+    .SetGenerator<int>((outcome, args) => TimeSpan.FromSeconds(args.Attempt)) // called for int result type only
+    .SetVoidGenerator((outcome, args) => TimeSpan.FromSeconds(args.Attempt)) // called for void result type only
+    .SetVoidGenerator(async (outcome, args) => await GenerateDelayAsync(out, args));
 ```
 
-In the preceding sample you see that `ShouldRetry` handles the following scenarios:
+The code above demonstrates how `OutcomeGenerator<RetryDelayArguments, TimeSpan>` can be used to register multiple synchronous and asynchronous generators with different signature under a single unified API.
+
+### Performance
+
+The callback API is non-allocating and fast. However, the performance varies depending on number of callbacks you register. Maximum performance is archived when your callbacks handles only a single result type with a single registered callback. In that scenario the underlying handler does not do any dictionary lookups.
+
+### Registering your custom callbacks
 
-- Asynchronous predicates;
-- Synchronous predicates;
-- Concrete value results;
-- Custom function-based callbacks;
-- Different result types;
-- Exception types or exception-based predicates;
+If you are registering asynchronous callbacks make sure that you respect the value of `ResilienceContext.IsSynchronous` property and execute you callbacks synchronously for synchronous executions. You should also use the `ResilienceContext.ContinueOnCapturedContext` in case your user code uses execution and synchronization context (i.e. asynchronous calls in UI applications).
```

Comment:
```suggestion
If you are registering asynchronous callbacks, make sure that you respect the value of the `ResilienceContext.IsSynchronous` property and execute you callbacks synchronously for synchronous executions. You should also use the `ResilienceContext.ContinueOnCapturedContext` in case your user code uses execution and synchronization context (i.e. asynchronous calls in UI applications, such as in Windows Forms or WPF applications).
```